Added modification to ProGuard setting (#1630)
OWASP/owasp-mstg: Document/0x05d-Testing-Data-Storage.md

While preparing the production release, you can use tools like `ProGuard` (included in Android Studio). [ProGuard](https://www.guardsquare.com/en/products/proguard "ProGuard") is a free Java class file shrinker, optimizer, obfuscator, and preverifier. It detects and removes unused classes, fields, methods, and attributes and can also be used to delete logging-related code.

> To determine whether all the `android.util.Log` class' logging functions have been removed, check the ProGuard configuration file (_proguard-rules.pro_) for the following options (according to [example of removing logging code](https://www.guardsquare.com/en/products/proguard/manual/examples#logging "ProGuard\'s exmaple of removing logging code") and [enabling ProGuard in Android Studio project ](https://developer.android.com/studio/build/shrink-code#enable "Android Developer - Enable shrinking, obfuscation, and optimization")):

Reviewer: ```suggestion
To determine whether all logging functions from the `android.util.Log` class have been removed, check the ProGuard configuration file (proguard-rules.pro) for the following options (according to this [example of removing logging code](https://www.guardsquare.com/en/products/proguard/manual/examples#logging "ProGuard\'s exmaple of removing logging code") and this article about [enabling ProGuard in an Android Studio project](https://developer.android.com/studio/build/shrink-code#enable "Android Developer - Enable shrinking, obfuscation, and optimization")):
```

Note that the example above only ensures that calls to the Log class' methods will be removed. If the string that will be logged is dynamically constructed, the code that constructs the string may remain in the bytecode. For example, the following code issues an implicit `StringBuilder` to construct the log statement:

```java

> Log.v("Private key tag","Private key [byte format]: " + key);

Reviewer: ```suggestion
Log.v("Private key tag", "Private key [byte format]: " + key);
```

```

The compiled bytecode, however, is equivalent to the bytecode of the following log statement, which constructs the string explicitly:

```java

> Log.v(new StringBuilder("Private key tag"),new StringBuilder("Private key [byte format]: ").append(key.toString()).toString());

Reviewer: ```suggestion
Log.v("Private key tag", new StringBuilder("Private key [byte format]: ").append(key.toString()).toString());
```

Reviewer: Why using stringbuilder here and not in the line above? Or vice versa

Author: Updating could be tricky. I think probably that there could be some leftover in bytecode, which is explained in that paragraph. Check again - tag is not in decompiled code. However log message is, example:
```
 public void sendMessage(View paramView) {
    Intent intent = new Intent((Context)this, DisplayMessageActivity.class);
    EditText editText = (EditText)findViewById(2131165238);
    String str = editText.getText().toString();
    StringBuilder stringBuilder = new StringBuilder("Sending message:");
    stringBuilder.append(editText.toString());
    stringBuilder.toString();
    intent.putExtra("com.example.myfirstapp.MESSAGE", str);
    startActivity(intent);
  }
```
In that case your fix is perfect :) Thanks for spotting.

Identify all defined `<provider>` elements:

```xml

> <provider  

Reviewer: ```suggestion
<provider
```

>       android:authorities="com.mwr.example.sieve.DBContentProvider"  

Reviewer: ```suggestion
      android:authorities="com.mwr.example.sieve.DBContentProvider"
```

>       android:exported="true"  

Reviewer: ```suggestion
      android:exported="true"
```

>       android:multiprocess="true"  

Reviewer: ```suggestion
      android:multiprocess="true"
```

      android:name=".DBContentProvider">

>     <path-permission  

Reviewer: ```suggestion
    <path-permission
```

>           android:path="/Keys"  

Reviewer: ```suggestion
          android:path="/Keys"
```

          android:readPermission="com.mwr.example.sieve.READ_KEYS"
          android:writePermission="com.mwr.example.sieve.WRITE_KEYS"
     />
</provider>

> <provider  

Reviewer: ```suggestion
<provider
```

>       android:authorities="com.mwr.example.sieve.FileBackupProvider"  

Reviewer: ```suggestion
      android:authorities="com.mwr.example.sieve.FileBackupProvider"
```

>       android:exported="true"  

Reviewer: ```suggestion
      android:exported="true"
```

>       android:multiprocess="true"  

Reviewer: ```suggestion
      android:multiprocess="true"
```
